dace/transformation/interstate/transient_reuse.py

```diff
@@ -0,0 +1,175 @@
+from dace import registry, sdfg as sd
+from dace.graph import nodes
+from dace.transformation import pattern_matching
+from dace.properties import make_properties
+import networkx as nx
+
+
+@registry.autoregister
+@make_properties
+class TransientReuse(pattern_matching.Transformation):
+    """ Implements the TransientReuse transformation.
+        Finds all possible reuses of arrays,
+        decides for a valid combination and changes sdfg accordingly.
+    """
+
+    @staticmethod
+    def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
+        return True
+
+    @staticmethod
+    def match_to_str(graph, candidate):
+        return graph.label
+
+    @staticmethod
+    def expressions():
+        return [sd.SDFG('_')]
+
+    def expansion(node):
+        pass
+
+    def apply(self, sdfg):
+
+        memory_before = 0
+        arrays = {}
+        for a in sdfg.arrays:
+            if sdfg.arrays[a].transient:
+                arrays[a] = 0
+                memory_before += sdfg.arrays[a].total_size
+
+        # only consider transients appearing in one single state
+        for state in sdfg.states():
+            for a in state.all_transients():
+                arrays[a] += 1
+
+        transients = set()
+        for a in arrays:
+            if arrays[a] == 1:
+                transients.add(a)
+
+        for state in sdfg.nodes():
+            # Copy the whole graph
+            G = nx.MultiDiGraph()
+            for n in state.nodes():
+                G.add_node(n)
+            for n in state.nodes():
+                for e in state.all_edges(n):
+                    G.add_edge(e.src, e.dst)
+
+            # Collapse all mappings and their scopes into one node
+            scope_dict = state.scope_dict(node_to_children=True)
+            for n in scope_dict:
+                if n is not None:
```

Comment:
Are you sure this works with nested scopes (map in a map)? I would simply loop over `scope_dict[None]` looking for `EntryNode`s to be safe.

Reply:
Running it works on e.g. lenet it finds all nested scopes... but if you don't trust the method I will change it.

Reply:
I don't trust it because the order of collapsing might leave uncollapsed top-level maps.

Reply:
Ok! Changed it yesterday ^^